File: apps/api/app/schemas/billing.py
from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class PlanOut(BaseModel):
    code: str
    tier: str
    price_som: int
    duration_days: int
    seats: int


class PlansOut(BaseModel):
    plans: List[PlanOut]


class CoinPackOut(BaseModel):
    code: str
    coins: int
    price_som: int


class CoinPacksOut(BaseModel):
    packs: List[CoinPackOut]


class CoinTransactionOut(BaseModel):
    delta: int
    reason: str
    balance_after: int
    created_at: datetime


class WalletOut(BaseModel):
    balance: int
    recent_transactions: List[CoinTransactionOut]


class BillingStatusOut(BaseModel):
    checkout_enabled: bool
    sandbox_enabled: bool
    providers: Dict[str, bool]
    family_plan_available: bool = False


class SubscriptionOut(BaseModel):
    is_premium: bool
    plan_code: Optional[str] = None
    status: Optional[str] = None
    provider: Optional[str] = None
    expires_at: Optional[datetime] = None
    seats: int = 1
    auto_renew: bool = False
    cancelled_at: Optional[datetime] = None


class CheckoutRequest(BaseModel):
    # Accepts either a sellable Plan.code or a coin-pack code (e.g.
    # "coins_small") — see api/v1/payments.py's _get_purchasable.
    plan_code: str = Field(min_length=1, max_length=24)
    provider: Literal["payme", "click", "uzum"]
    return_url: Optional[str] = Field(default=None, max_length=500)


class CheckoutOut(BaseModel):
    order_id: str
    checkout_url: str
    amount_som: int


class SandboxActivateRequest(BaseModel):
    plan_code: str = Field(min_length=1, max_length=24)


class ReferralOut(BaseModel):
    code: str
    invited: int
    rewarded: int
    reward_days: int


class MessageOut(BaseModel):
    message: str

File: apps/web/lib/billing.ts
import { apiFetch } from "@/lib/api";

export interface Plan {
  code: string;
  tier: string;
  price_som: number;
  duration_days: number;
  seats: number;
}

export interface Subscription {
  is_premium: boolean;
  plan_code: string | null;
  status: string | null;
  provider: string | null;
  expires_at: string | null;
  seats: number;
  auto_renew: boolean;
  cancelled_at: string | null;
  show_welcome: boolean;
}

export interface Checkout {
  order_id: string;
  checkout_url: string;
  amount_som: number;
}

export type PaymentProvider = "payme" | "click" | "uzum";

export interface BillingStatus {
  checkout_enabled: boolean;
  sandbox_enabled: boolean;
  providers: {
    payme: boolean;
    click: boolean;
    uzum: boolean;
  };
  family_plan_available: boolean;
}

export interface ReferralInfo {
  code: string;
  invited: number;
  rewarded: number;
  reward_days: number;
}

export interface ManualPayment {
  reference: string;
  plan_code: string;
  amount_som: number;
  status: string;
  created_at: string;
}

export const billingApi = {
  /** Records that the learner says they transferred the price to the card.
   *  Grants nothing — staff activate by hand — but gives them a reference
   *  and gives us a queue. */
  requestManualPayment: (planCode: string) =>
    apiFetch<ManualPayment>("/billing/manual-payment", {
      method: "POST",
      body: { plan_code: planCode },
      auth: true,
    }),

  manualPayment: () =>
    apiFetch<ManualPayment | null>("/billing/manual-payment", { auth: true }),

  plans: () => apiFetch<{ plans: Plan[] }>("/billing/plans"),

  status: () => apiFetch<BillingStatus>("/billing/status"),

  subscription: () => apiFetch<Subscription>("/billing/subscription", { auth: true }),

  /** Marks the "Premium is on" message as seen, so it appears once per
   *  activation rather than on every page load. */
  markWelcomed: () =>
    apiFetch<{ message: string }>("/billing/subscription/welcomed", {
      method: "POST",
      auth: true,
    }),

  checkout: (planCode: string, provider: PaymentProvider, returnUrl: string, idempotencyKey?: string) =>
    apiFetch<Checkout>("/billing/checkout", {
      method: "POST",
      body: { plan_code: planCode, provider, return_url: returnUrl },
      auth: true,
      headers: idempotencyKey ? { "Idempotency-Key": idempotencyKey } : undefined,
    }),

  sandboxActivate: (planCode: string) =>
    apiFetch<Subscription>("/billing/sandbox-activate", {
      method: "POST",
      body: { plan_code: planCode },
      auth: true,
    }),

  cancel: () =>
    apiFetch<{ message: string }>("/billing/cancel", { method: "POST", auth: true }),

  referral: () => apiFetch<ReferralInfo>("/billing/referral", { auth: true }),
};

export function formatSom(amount: number): string {
  return amount.toLocaleString("ru-RU");
}
